sentry-core/src/main/java/io/sentry/core/cache/DiskCache.java

```diff
@@ -0,0 +1,152 @@
+      return;
+    }
+
+    File eventFile = getEventFile(event);
+    if (eventFile.exists()) {
+      logIfNotNull(options.getLogger(), DEBUG,
+        "Not adding Event to offline storage because it already exists: %s", eventFile.getAbsolutePath());
+      return;
+    } else {
+      logIfNotNull(options.getLogger(), DEBUG, "Adding Event to offline storage: %s", eventFile.getAbsolutePath());
+    }
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(eventFile);
+         Writer wrt = new OutputStreamWriter(fileOutputStream, UTF8)) {
+      serializer.serialize(event, wrt);
+    } catch (Exception e) {
+      logIfNotNull(options.getLogger(), ERROR, "Error writing Event to offline storage: %s", event.getEventId());
+    }
+  }
+
+  @Override
+  public void discard(SentryEvent event) {
+    File eventFile = getEventFile(event);
+        if (eventFile.exists()) {
+            logIfNotNull(options.getLogger(), DEBUG, "Discarding event from cache: %s",
+              eventFile.getAbsolutePath());
+
+            if (!eventFile.delete()) {
+                logIfNotNull(options.getLogger(), WARNING, "Failed to delete Event: %s", eventFile.getAbsolutePath());
+            }
+        } else {
+          logIfNotNull(options.getLogger(), DEBUG, "Event was not cached: %s", eventFile.getAbsolutePath());
+        }
+  }
+
+  private int getNumberOfStoredEvents() {
+    return allEventFiles().length;
+  }
+
+  private void checkDirectoryValid() {
+    if (!directory.isDirectory() || !directory.canWrite() || !directory.canRead()) {
+      throw new IllegalStateException("The directory for caching Sentry events is inaccessible.");
+    }
+  }
+
+  private File getEventFile(SentryEvent event) {
+    return new File(directory.getAbsolutePath(), event.getEventId().toString() + FILE_SUFFIX);
+  }
+
+  @NotNull
+  @Override
+  public Iterator<SentryEvent> iterator() {
+    File[] allCachedEvents = allEventFiles();
+
+    List<SentryEvent> ret = new ArrayList<>(allCachedEvents.length);
+
+    char[] buffer = new char[1024];
+    for (File f : allCachedEvents) {
+      try (InputStreamReader rdr = new InputStreamReader(new BufferedInputStream(new FileInputStream(f)), UTF8)) {
+
+        int cnt;
+        StringBuilder bld = new StringBuilder();
+        while ((cnt = rdr.read(buffer)) != -1) {
+          bld.append(buffer, 0, cnt);
+        }
+
+        ret.add(serializer.deserializeEvent(bld.toString()));
+      } catch (FileNotFoundException e) {
+        logIfNotNull(options.getLogger(), DEBUG, "Event file '%s' disappeared while converting all cached files to events.",
+          f.getAbsolutePath());
+      } catch (IOException e) {
+        logIfNotNull(options.getLogger(), ERROR,
+          format("Error while reading cached event from file %s", f.getAbsolutePath()), e);
+      }
+    }
+
+    return ret.iterator();
+  }
+
+  private File[] allEventFiles() {
+    checkDirectoryValid();
```

Comment:
Do we need to run this before each access? Since this runs in the constructor (which means I/O in the ctor) can we get away with just doing that in the ctor?
Is the goal to report early if some other process changed the directory underneath us? Or disk is full or else?
I dunno if each of those does an I/O call but sounds like we'd be:
> isDirectory,canRead,canWrite,fileExists,fileWrite

Are these 6 I/O calls for each write?
And do we do the same for delete except the last one is _fileDelete instead of write_?

Not saying it's a problem but if the goal is to have better logging, maybe we only run the checks if logging is enabled? If there's no logging just attempt the operation we intended and .. I guess can't log the error.